### PR TITLE
Adding AMD MI200/300 support

### PR DESCRIPTION
**Pull request by rafsalas19**
- Add ROCM Dockerfile
- Modifications to build and run with a staged ROCM docker image
- Modify VM meta data collection and printing. Will work with run script launch and docker only launch
- Add RCCL test
- Addition of a time threshold to the XID error check. The check will discard any xid errors that occure before the time threshold. Default is 2 hours.
- Other bug fixes

### CODE REVIEW

File: conf/nd96isr_mi300x_v5.conf [hunk 1]
```diff
@@ -48,6 +48,7 @@
 ####
   * || check_gpu_count 8 rocm
   * || check_gpu_bw 48 44
+  * || check_rccl_allreduce 304 1 16G
```
Comment (moes1) [inline conf/nd96isr_mi300x_v5.conf:51]: How tight do we need to be to the observed value? Does 304 have some buffer added?

Comment (rafsalas19) [reply conf/nd96isr_mi300x_v5.conf:51]: We have about a 5% buffer. 320 was what I observed normally. This PR was also tested at scale on an MI300 cluster.

File: dockerfile/build_image.sh [hunk 2]
```diff
@@ -96,8 +96,8 @@ if [[ "$build_type" == "cuda" ]]; then
     popd
 elif [[ "$build_type" == "rocm" ]]; then
     echo "AMD runtime selected"
-    echo not supported yet but coming soon
-    exit 1
+    IMAGE="azurenodehealthchecks.azurecr.io/staging/aznhc/aznhc-rocm"
```
Comment (rajvi) [inline dockerfile/build_image.sh:99]: Will we have to build this image for validation tests as well along with "mcr.microsoft.com/aznhc/aznhc-nv"?

**Review by rajvi — APPROVED** [review]

File: customTests/azure_common.nhc [hunk 3]
```diff
@@ -45,7 +46,15 @@ function background_cleanup() {
...
+
+   # Default to Nvidia NCCL-Test
+   REDUCE_PATH=$ALL_REDUCE_PATH
+   if lspci | grep -iq AMD ; then
+      # Check if AMD RCCL-Test is installed
+      REDUCE_PATH=$RCCL_ALL_REDUCE_PATH 
+   fi
+
+   if [[ ! -f $REDUCE_PATH ]]; then
       die 1 "$FUNCNAME: all reduce executable path: $ALL_REDUCE_PATH not found. Ensure nccl-tests is installed and built in. FaultCode: NHCNA"
```
Comment (yakovdyadkin) [inline customTests/azure_common.nhc]: Should the die message use `REDUCE_PATH`?

File: customTests/azure_ib_write_bw_gdr.nhc [hunk 4]
```diff
@@ -62,7 +62,7 @@ function run_ib_bw_gdr(){
   dbg "ib_write_lb_${device}: $ib_bandwidth Gbps"
   if (( $(echo "$ib_bandwidth < $EXP_IB_BW" | bc -l) )); then
 	 log "$IB_WRITE_BW_OUT2"
-    die 1 "$FUNCNAME: $IB_WRITE_BW, IB=$device, $device_peer, IB BW (expected > $EXP_IB_BW Gbps, but measured $ib_bandwidth Gbps. FaultCode: NHC2003"
+    die 1 "check_ib_bw_gdr: $IB_WRITE_BW, IB=$device, $device_peer, IB BW (expected > $EXP_IB_BW Gbps, but measured $ib_bandwidth Gbps. FaultCode: NHC2003"
```
Comment (yakovdyadkin) [inline customTests/azure_ib_write_bw_gdr.nhc:65]: I assume this is to make the die message consistent for this NHC test as this func is wrapped by `check_ib_bw_gdr`. Just want to make sure there is no parser that could have looked for `run_ib_bw_gdr`

File: customTests/azure_ib_write_bw_non_gdr.nhc [hunk 5]
```diff
@@ -77,7 +77,7 @@ function ib_write(){
          return 1
    fi
    dbg "ib_write_lb_${device}: $ib_bandwidth Gbps"
-   pass 0 "$FUNCNAME: IB write bandwidth non gdr test passed"
+   pass 0 "check_ib_bw_non_gdr: IB write bandwidth non gdr test passed for IB=$device, IB BW=$ib_bandwidth Gbps"
```
Comment (yakovdyadkin) [inline customTests/azure_ib_write_bw_non_gdr.nhc:80]: Same as my other comment related to die message.

File: dockerfile/aznhc-entrypoint.sh [hunk 6]
```diff
@@ -1,5 +1,13 @@
 #! /bin/bash
 
+function collect_meta_data(){
+    # get meta data for VM and underlying host
+    vmhostname=$(hostname)
+    vmid=$( curl -H Metadata:true --max-time 10 -s  "http://169.254.169.254/metadata/instance/compute/vmId?api-version=2021-03-01&format=text")
+    vm_name=$(curl -H Metadata:true --max-time 10 -s "http://169.254.169.254/metadata/instance/compute/name?api-version=2021-11-15&format=text")
```
Comment (yakovdyadkin) [inline dockerfile/aznhc-entrypoint.sh]: Nitpick: is there a reason `vm_name` uses an underscore?

Comment (rafsalas19) [reply dockerfile/aznhc-entrypoint.sh]: not really, let me handle

File: customTests/azure_ib_write_bw_gdr.nhc (hunk 4 above)
Comment (rafsalas19) [reply customTests/azure_ib_write_bw_gdr.nhc:65]: Yea this is more so to have the function name reported be the same as the check's name (which would be the initial function NHC enters)

File: dockerfile/build_image.sh (hunk 2 above)
Comment (rafsalas19) [reply dockerfile/build_image.sh:99]: Maybe. Not a high priority at the moment

File: customTests/azure_common.nhc (hunk 3 above)
Comment (rafsalas19) [reply customTests/azure_common.nhc]: Addressed

File: customTests/azure_ib_write_bw_gdr.nhc (hunk 4 above)
Comment (rafsalas19) [reply customTests/azure_ib_write_bw_gdr.nhc:65]: Addressed

File: dockerfile/aznhc-entrypoint.sh (hunk 6 above)
Comment (rafsalas19) [reply dockerfile/aznhc-entrypoint.sh]: Addressed

File: customTests/azure_ib_write_bw_non_gdr.nhc (hunk 5 above)
Comment (rafsalas19) [reply customTests/azure_ib_write_bw_non_gdr.nhc:80]: Addressed

**Pull request merged**